File: minigit-3/ts/src/commands/cat-file.ts
// cat-file command - Examine objects

import { findRepoRoot } from '../utils';
import { resolveRef } from '../refs';
import { readObject, parseCommit, parseTree, parseTag, getObjectType, getObjectSize, prettyPrintObject } from '../objects';

export function catFile(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  let showType = false;
  let showSize = false;
  let prettyPrint = false;
  let objectType: string | null = null;
  let objectRef: string | null = null;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-t') {
      showType = true;
    } else if (arg === '-s') {
      showSize = true;
    } else if (arg === '-p') {
      prettyPrint = true;
    } else if (arg === 'blob' || arg === 'tree' || arg === 'commit' || arg === 'tag') {
      if (!objectType) {
        objectType = arg;
      } else {
        objectRef = arg;
      }
    } else if (!arg.startsWith('-')) {
      objectRef = arg;
    }
  }

  if (!objectRef) {
    console.error('fatal: object required');
    process.exit(1);
  }

  // Handle path syntax: <commit>:<path>
  if (objectRef.includes(':')) {
    const sha = resolvePath(objectRef, repoRoot);
    if (!sha) {
      console.error(`fatal: bad object ${objectRef}`);
      process.exit(1);
    }
    objectRef = sha;
  }

  // Handle special suffixes like HEAD^{tree}
  const sha = resolveRef(objectRef, repoRoot);
  if (!sha) {
    console.error(`fatal: Not a valid object name ${objectRef}`);
    process.exit(1);
  }

  try {
    if (showType) {
      const type = getObjectType(sha, repoRoot);
      console.log(type);
    } else if (showSize) {
      const size = getObjectSize(sha, repoRoot);
      console.log(size);
    } else if (prettyPrint) {
      const output = prettyPrintObject(sha, repoRoot);
      console.log(output);
    } else if (objectType) {
      // cat-file <type> <object> - verify type and show content
      const obj = readObject(sha, repoRoot);
      if (obj.type !== objectType) {
        console.error(`fatal: git cat-file: ${objectRef} is not a ${objectType}`);
        process.exit(1);
      }
      process.stdout.write(obj.content);
    } else {
      // Default: pretty print
      const output = prettyPrintObject(sha, repoRoot);
      console.log(output);
    }
  } catch (e) {
    console.error(`fatal: git cat-file: ${objectRef}: cannot read object`);
    process.exit(1);
  }
}

function resolvePath(ref: string, repoRoot: string): string | null {
  const colonIndex = ref.indexOf(':');
  const commitRef = ref.slice(0, colonIndex);
  const pathPart = ref.slice(colonIndex + 1);

  const sha = resolveRef(commitRef, repoRoot);
  if (!sha) return null;

  const obj = readObject(sha, repoRoot);
  if (obj.type !== 'commit') return null;

  const commit = parseCommit(obj.content);

  // Navigate tree to find path
  let currentTree = commit.tree;
  const parts = pathPart.split('/').filter(p => p);

  for (let i = 0; i < parts.length; i++) {
    const part = parts[i];
    const treeObj = readObject(currentTree, repoRoot);
    const entries = parseTree(treeObj.content);

    const entry = entries.find(e => e.name === part);
    if (!entry) return null;

    if (i === parts.length - 1) {
      return entry.sha;
    }

    if (entry.type !== 'tree') return null;
    currentTree = entry.sha;
  }

  return currentTree;
}

File: minigit-3/py/commands/cat_file.py
"""cat-file command - Examine objects."""

import sys
from pathlib import Path


def run(args: list[str]) -> int:
    """Examine object internals."""
    from utils import find_repo_root, read_object
    from refs import resolve_ref
    from objects import read_commit, read_tree, read_tag

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    show_type = False
    show_size = False
    pretty_print = False
    obj_type_filter = None
    target = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-t':
            show_type = True
        elif arg == '-s':
            show_size = True
        elif arg == '-p':
            pretty_print = True
        elif arg in ('blob', 'tree', 'commit', 'tag'):
            obj_type_filter = arg
            if i + 1 < len(args):
                target = args[i + 1]
                i += 1
        elif not arg.startswith('-'):
            target = arg
        i += 1

    if target is None:
        print("error: object required", file=sys.stderr)
        return 1

    # Handle path specifier (e.g., HEAD:file.txt)
    if ':' in target:
        ref_part, path = target.split(':', 1)
        ref_sha = resolve_ref(repo_root, ref_part)
        if ref_sha is None:
            print(f"error: unknown revision '{ref_part}'", file=sys.stderr)
            return 1

        # Get the blob SHA for the file
        commit = read_commit(repo_root, ref_sha)
        entry = get_tree_entry(repo_root, commit['tree'], path)
        if entry is None:
            print(f"error: path '{path}' not found in '{ref_part}'", file=sys.stderr)
            return 1
        target = entry.sha

    # Handle ^{tree} specifier
    elif '^{' in target:
        base, spec = target.split('^{')
        spec = spec.rstrip('}')
        sha = resolve_ref(repo_root, base)
        if sha is None:
            print(f"error: unknown revision '{base}'", file=sys.stderr)
            return 1

        if spec == 'tree':
            commit = read_commit(repo_root, sha)
            target = commit['tree']
        elif spec == 'commit':
            target = sha
        else:
            target = sha
    else:
        # Resolve the reference
        sha = resolve_ref(repo_root, target)
        if sha is None:
            print(f"error: not a valid object name '{target}'", file=sys.stderr)
            return 1
        target = sha

    # Read object
    try:
        obj_type, data = read_object(repo_root, target)
    except Exception as e:
        print(f"error: object '{target}' not found", file=sys.stderr)
        return 1

    # Check type filter
    if obj_type_filter and obj_type != obj_type_filter:
        print(f"error: object type mismatch: expected {obj_type_filter}, got {obj_type}", file=sys.stderr)
        return 1

    if show_type:
        print(obj_type)
    elif show_size:
        print(len(data))
    elif pretty_print or obj_type_filter:
        if obj_type == 'blob':
            try:
                print(data.decode(), end='')
            except:
                sys.stdout.buffer.write(data)
        elif obj_type == 'tree':
            entries = read_tree(repo_root, target)
            for entry in entries:
                type_name = 'tree' if entry.mode == '40000' else 'blob'
                print(f"{entry.mode} {type_name} {entry.sha}\t{entry.name}")
        elif obj_type == 'commit':
            commit = read_commit(repo_root, target)
            print(f"tree {commit['tree']}")
            for parent in commit['parents']:
                print(f"parent {parent}")
            print(f"author {commit['author']}")
            print(f"committer {commit['committer']}")
            print()
            print(commit['message'])
        elif obj_type == 'tag':
            tag = read_tag(repo_root, target)
            print(f"object {tag['object']}")
            print(f"type {tag['type']}")
            print(f"tag {tag['tag']}")
            print(f"tagger {tag['tagger']}")
            print()
            print(tag['message'])
    else:
        # Raw output
        try:
            print(data.decode(), end='')
        except:
            sys.stdout.buffer.write(data)

    return 0


def get_tree_entry(repo_root, tree_sha: str, path: str):
    """Get entry from tree by path."""
    from objects import read_tree

    parts = path.split('/')
    current_sha = tree_sha

    for i, part in enumerate(parts):
        entries = read_tree(repo_root, current_sha)
        found = None
        for entry in entries:
            if entry.name == part:
                found = entry
                break

        if found is None:
            return None

        if i < len(parts) - 1:
            if found.mode != '40000':
                return None
            current_sha = found.sha
        else:
            return found

    return None
